src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs

```diff
@@ -67,434 +113,658 @@ public JsonWebToken(string header, string payload)
+#endif
+                }
+                catch (Exception ex)
+                {
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14101, encodedJson.Substring(Dot2, Dot2 - Dot1), encodedJson), ex));
+                }
             }
-            catch (Exception ex)
+            else
             {
-                throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14102, tokenParts[0], rawData), ex));
+                // empty payload for JWE's {encrypted tokens}.
+#if NET45
+                Payload = new JsonClaimSet("{}");
+#else
+                Payload = new JsonClaimSet(JsonDocument.Parse("{}"));
+#endif
+                if (Dot3 == encodedJson.Length)
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14100, encodedJson)));
+
+                Dot4 = encodedJson.IndexOf('.', Dot3 + 1);
+
+                // JWE needs to have 4 dots
+                if (Dot4 == -1 || Dot4 == encodedJson.Length)
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14100, encodedJson)));
+
+                // too many dots...
+                if (encodedJson.IndexOf('.', Dot4 + 1) != -1)
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14100, encodedJson)));
+
+                // right number of dots for JWE
+                _hChars = encodedJson.ToCharArray(0, Dot1);
+
+                // header cannot be empty
+                if (_hChars.Length == 0)
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14307, encodedJson)));
+
+                HeaderAsciiBytes = Encoding.ASCII.GetBytes(_hChars);
+                try
+                {
+                    Header = new JsonClaimSet(Base64UrlEncoder.UnsafeDecode(_hChars));
+                }
+                catch (Exception ex)
+                {
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14102, encodedJson.Substring(0, Dot1), encodedJson), ex));
+                }
+
+                // dir does not have any key bytes
+                char[] encryptedKeyBytes = encodedJson.ToCharArray(Dot1 + 1, Dot2 - Dot1 - 1);
+                if (encryptedKeyBytes.Length != 0)
+                {
+                    EncryptedKeyBytes = Base64UrlEncoder.UnsafeDecode(encryptedKeyBytes);
+                    _encryptedKey = encodedJson.Substring(Dot1 + 1, Dot2 - Dot1 - 1);
+                }
+                else
+                {
+                    _encryptedKey = string.Empty;
+                }
+
+                char[] ivChars = encodedJson.ToCharArray(Dot2 + 1, Dot3 - Dot2 - 1);
```

Comment:
>ivChars [](http://example.com/codeflow?start=23&length=7)

super nit: ivChars is a rough name since it sort of seems like IV=4 so 4th part, at least that's where my mind first went. something like initVectorChars is only slightly longer and much clearer.